packages/core/src/utils/transactions.ts

```diff
@@ -0,0 +1,312 @@
+      console.warn(
+        `${from} already approved ${approveSpender}. In some tokens, approval will fail unless the allowance is reset to 0 before re-approving again.`
+      )
+    }
+
+    const erc20 = new ethers.Interface(erc20ABI)
+
+    const tokenApproveTransaction = {
+      // TODO: should we include transaction options?
+      from,
+      to: tokenAddress,
+      data: erc20.functions.approve.encode([approveSpender, tokenValue]),
+    }
+
+    delete transaction.token
+
+    return {
+      ...transaction,
+      pretransaction: tokenApproveTransaction,
+    }
+  }
+
+  return transaction
+}
+
+export async function applyForwardingFeePretransaction(
+  forwardingTransaction: TransactionRequestData,
+  provider: ethers.providers.Provider
+) {
+  const { to: forwarderAddress, from } = forwardingTransaction
+
+  const forwarderFee = new ethers.Contract(
+    forwarderAddress,
+    forwarderFeeAbi,
+    provider
+  )
+
+  const feeDetails = { amount: ethers.constants.Zero, tokenAddress: '' }
+  try {
+    const overrides = {
+      from,
+    }
+    // Passing the EOA as `msg.sender` to the forwardFee call is useful for use cases where the fee differs relative to the account
+    const returnValues = await forwarderFee.forwardFee(overrides) // forwardFee() returns (address, uint256)
+    feeDetails.tokenAddress = returnValues.tokenAddress
+    feeDetails.amount = ethers.utils.bigNumberify(returnValues.amount)
+  } catch (err) {
+    // Not all forwarders implement the `forwardFee()` interface
+  }
+
+  if (feeDetails.tokenAddress && feeDetails.amount.gt(ethers.constants.Zero)) {
+    // Needs a token approval pretransaction
+    const forwardingTxWithTokenData: transactionWithTokeData = {
+      ...forwardingTransaction,
+      token: {
+        address: feeDetails.tokenAddress,
+        spender: forwarderAddress, // since it's a forwarding transaction, always show the real spender
+        value: feeDetails.amount.toString(),
+      },
+    }
+    return applyPretransaction(forwardingTxWithTokenData, provider)
+  }
+
+  return forwardingTransaction
+}
+
+export async function getRecommendedGasLimit(
+  estimatedGasLimit: ethers.types.BigNumber,
+  provider: ethers.providers.Provider,
+  { gasFuzzFactor = DEFAULT_GAS_FUZZ_FACTOR } = {}
+) {
+  const latestBlockNumber = await provider.getBlockNumber()
+  const latestBlock = await provider.getBlock(latestBlockNumber)
+  const latestBlockGasLimit = latestBlock.gasLimit
+
+  const upperGasLimit = latestBlockGasLimit.mul(
+    ethers.utils.bigNumberify(PREVIOUS_BLOCK_GAS_LIMIT_FACTOR)
+  )
+
+  const bufferedGasLimit = estimatedGasLimit.mul(
+    ethers.utils.bigNumberify(gasFuzzFactor)
+  )
+
+  if (estimatedGasLimit.gt(upperGasLimit)) {
+    // TODO: Consider whether we should throw an error rather than returning with a high gas limit
+    return estimatedGasLimit
+  } else if (bufferedGasLimit.lt(upperGasLimit)) {
+    return bufferedGasLimit
+  } else {
+    return upperGasLimit
+  }
```

Comment:
Nitpick: we could flatten this a bit:

```suggestion
  if (estimatedGasLimit.gt(upperGasLimit)) {
    // TODO: Consider whether we should throw an error rather than returning with a high gas limit
    return estimatedGasLimit
  }
  if (bufferedGasLimit.lt(upperGasLimit)) {
    return bufferedGasLimit
  }
  return upperGasLimit
```